# Практическая работа №4: Элементы корреляционного анализа. Проверка статистической гипотезы о равенстве коэффициента корреляции нулю

Выполнили студентки гр. 2383 Щёголева Надежда и Семененко Анна. Вариант №28

## Цель работы

Освоение основных понятий, связанных с корреляционной зависимостью между случайными величинами, статистическими гипотезами и проверкой их «справедливости».

## Основные теоретические положения



Выборочный коэффициент корреляции:

$$\bar{r}_{xy} = \frac{\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij}y_i x_j - N\bar{x}_в \bar{y}_в}{N S_x S_y}$$

Вычисление выборочного коэффициента корреляции с помощью условных вариант:

$$\vec{r}_{xy} = \frac{\sum_{i=1}^{K_y} \sum_{j=1}^{K_x} n_{ij} u_i v_j - N \bar{u}_в \bar{v}_в}{N \bar{\sigma}_u \bar{\sigma}_v}$$


$$u_i = \frac{x_i - C_x}{h_x}, \quad v_j = \frac{y_j - C_y}{h_y}$$

$$\bar{\sigma}_u = \sqrt{\overline{u^2} - \bar{u}^2}, \quad \bar{\sigma}_v = \sqrt{\overline{v^2} - \bar{v}^2}$$


$$\bar{u}_\sigma = \frac{\sum n_i u_i}{N}, \quad \bar{v}_\sigma = \frac{\sum n_j v_j}{N}$$


$$\overline{u^2} = \frac{\sum n_i u_i^2}{N}, \quad \overline{v^2} = \frac{\sum n_j v_j^2}{N}$$

Z-преобразование Фишера:

$$\overline{Z} = 0.5 \ln \frac{1 + \overline{r}_{xy}}{1 - \overline{r}_{xy}} = 1.15131 \lg \frac{1 + \overline{r}_{xy}}{1 - \overline{r}_{xy}}$$

СКВО:

$$\overline{\sigma}_z = \frac{1}{\sqrt{N-3}}$$

Доверительный интервал для Z-преобразованной величины

$$\left( \bar{z} - \lambda(\gamma) \sigma_z,\; \bar{z} + \lambda(\gamma) \sigma_z \right)$$

где $\lambda(\gamma)$ определяется из условия:
$$\Phi[\lambda(\gamma)] = \frac{\gamma}{2}$$

Для перехода от интервала для $Z$ к доверительному интервалу для коэффициента корреляции $r$:

$$r = \text{th}(z) = \frac{e^{2z} - 1}{e^{2z} + 1}$$

Проверка значимости коэффициента корреляции

$$T_{\text{набл}} = \frac{\overline{r}_{xy} \sqrt{N - 2}}{\sqrt{1 - \overline{r}_{xy}^2}}$$


Гипотеза $H_0$: коэффициент корреляции в генеральной совокупности равен нулю

Гипотеза $H_1$: коэффициент корреляции в генеральной совокупности значимо отличается от нуля

$\|T_{\text{набл}}\| \leq t_{\text{крит}}(\alpha, k)$ - Нет оснований отвергать $H_0$. Выборочный коэффициент вариации $\bar{r}_{xy}$ незначим

$\|T_{\text{набл}}\| > t_{\text{крит}}(\alpha, k)$ - $H_0$ отвергается. Выборочный коэффициент вариации $\bar{r}_{xy}$ значим.


## Постановка задачи

Из заданной генеральной совокупности сформировать выборку по второму признаку. Провести статистическую обработку второй выборки в объёме практических работ №1 и №2, с целью определения точечных статистических оценок параметров распределения исследуемого признака (математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии, эксцесса и коэффициента вариации). Для системы двух случайных величин X (первый признак) и Y (второй признак) сформировать двумерную выборку и найти статистическую оценку коэффициента корреляции, построить доверительный интервал для коэффициента корреляции и осуществить проверку статистической гипотезы о равенстве коэффициента корреляции нулю. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

In [ ]:
import pandas as pd
import math
import matplotlib.pyplot as plt
import numpy as np
df = pd.read_csv('penguins.csv')
N = 117
K = 7

 1. Определение точечных статистических оценок параметров распределения  обоих исследуемых признаков было проведено во второй практической работе:


| Статистический показатель | Длина клюва | Высота клюва |
|:--------------------------|:-----------:|:-------------:|
| **Математическое ожидание** $\bar{x}_в$ | 43.94 | 16.14 |
| **Дисперсия** $D_в$ | 22.91 | 4.22 |
| **СКО** $\sigma_в$ | 4.79 | 2.05 |
| **Коэффициент асимметрии** $\bar{A}_s$ | -0.01 | 0.53 |
| **Эксцесс** $\bar{E}_в$ | -0.6 | -0.87 |
| **Мода** $M_o^*$ | 46.74 | 14.76 |
| **Медиана** $M_e^*$ | 44.89 | 15.67 |
| **Коэффициент вариации** $V^*$ | 10.88% | 12.69% |


Длина клюва: $\bar{x} < M_e < M_o$ — левосторонняя незначительная асимметрия, плосковершинное распределение

Высота клюва: $\bar{x} > M_e > M_o$ — правосторонняя существенная асимметрия, плосковершинное распределение

Оба признака имеют небольшую вариацию < 30% , значит статистическая совокупность однородная.

2. Интервальные вариационные ряды также уже построены первой работе:

In [ ]:
df_lenth = pd.read_json('df_stats_len.json')
display(df_lenth)

,Интервал,Середина интервала,Абсолютная частота (m_i),Относительная частота (m_i/N),Накопленная абс. частота,Накопленная отн. частота
1,"[34.5, 37.9]",36.20,17,0.1453,17,0.1453
2,"[38.1, 41.6]",39.85,23,0.1966,40,0.3419
3,"[42.0, 45.2]",43.60,23,0.1966,63,0.5385
4,"[45.3, 48.7]",47.00,35,0.2991,98,0.8376
5,"[49.0, 50.7]",49.85,17,0.1453,115,0.9829
6,[54.3],54.30,1,0.0085,116,0.9915
7,[59.6],59.60,1,0.0085,117,1.0000


In [ ]:
df_depth = pd.read_json('df_stats_depth.json')
display(df_depth)

,Интервал,Середина интервала,Абсолютная частота (m_i),Относительная частота (m_i/N),Накопленная абс. частота,Накопленная отн. частота
1,"[13.1, 14.2]",13.65,23,0.1966,23,0.1966
2,"[14.3, 15.3]",14.80,31,0.2650,54,0.4615
3,"[15.4, 16.5]",15.95,19,0.1624,73,0.6239
4,"[16.6, 17.5]",17.05,11,0.0940,84,0.7179
5,"[17.7, 18.8]",18.25,17,0.1453,101,0.8632
6,"[18.9, 19.5]",19.20,10,0.0855,111,0.9487
7,"[20.0, 21.1]",20.55,6,0.0513,117,1.0000


3. По полученному двумерному интервальному вариационному ряду построитм корреляционную таблицу.

In [ ]:
length_bins = [34.5, 38.1, 42.0, 45.3, 49.0, 54.3, 59.6, 60.0]
length_labels = ['[34.5-37.9]', '[38.1-41.6]', '[42.0-45.2]',
                 '[45.3-48.7]', '[49.0-50.7]', '[54.3]', '[59.6]']

depth_bins = [13.1, 14.3, 15.4, 16.6, 17.7, 18.9, 20.0, 21.2]
depth_labels = ['[13.1-14.2]', '[14.3-15.3]', '[15.4-16.5]',
                '[16.6-17.5]', '[17.7-18.8]', '[18.9-19.5]', '[20.0-21.1]']

df_corr = df.copy()
df_corr['length_interval'] = pd.cut(df_corr['culmen_length_mm'],
                                     bins=length_bins,
                                     labels=length_labels,
                                     include_lowest=True,
                                     right=False)

df_corr['depth_interval'] = pd.cut(df_corr['culmen_depth_mm'],
                                    bins=depth_bins,
                                    labels=depth_labels,
                                    include_lowest=True,
                                    right=False)

corr_table = pd.crosstab(df_corr['length_interval'],
                          df_corr['depth_interval'],
                          margins=True,
                          margins_name='Всего')

print("Корреляционная таблица:")
display(corr_table)

Корреляционная таблица:


depth_interval,[13.1-14.2],[14.3-15.3],[15.4-16.5],[16.6-17.5],[17.7-18.8],[18.9-19.5],[20.0-21.1],Всего
length_interval,,,,,,,,
[34.5-37.9],0,0,2,4,7,3,1,17
[38.1-41.6],1,0,1,4,9,4,4,23
[42.0-45.2],11,5,2,1,1,3,0,23
[45.3-48.7],11,17,6,0,0,0,1,35
[49.0-50.7],0,9,7,1,0,0,0,17
[54.3],0,0,1,0,0,0,0,1
[59.6],0,0,0,1,0,0,0,1
Всего,23,31,19,11,17,10,6,117


Прослеживается обратная зависимость между высотой клюва и его длиной: чем больше высота клюва, тем меньше его длина (и наоборот). Это видно по распределению частот:

Для небольших высот ([13.1-14.2], [14.3-15.3]) характерна бóльшая длина ([42.0-45.2] ,[45.3-48.7]) - 44 особи

Для больших высот ([16.6-17.5],	[17.7-18.8],	[18.9-19.5]) преобладает меньшая длина ([34.5-37.9] , [38.1-41.6]) – 31 особь


4. Исходя из результатов корреляционной таблицы вычислим значение выборочного коэффициента корреляции двумя способами: с помощью стандартной формулы и с помощью условных вариант.

In [ ]:
x_B = 43.941
y_B = 16.144
S_x = 4.786
S_y = 2.055

In [ ]:
freq_matrix = corr_table.iloc[:-1, :-1].values
sum_xy = 0

for i in range(K):
    for j in range(K):
        n_ij = freq_matrix[i, j]
        if n_ij > 0:
          contribution = n_ij * df_lenth.loc[i+1]['Середина интервала'] * df_depth.loc[j+1]['Середина интервала']
          sum_xy += contribution
print(f"\nСумма ∑∑ n_ij * x_i * y_j = {sum_xy:.2f}")


Сумма ∑∑ n_ij * x_i * y_j = 82354.24


In [ ]:
r = (sum_xy - N * x_B * y_B) / (N * S_x * S_y)
print('Выборочный коэффициент корреляции: ', round(r, 2))

Выборочный коэффициент корреляции:  -0.56


Вычисление значение выборочного коэффициента корреляции с помощью условных вариант:

In [ ]:
data_l = {
    'i': [1, 2, 3, 4, 5, 6, 7],
    'x_i': [36.2, 39.85, 43.6, 47.0, 49.85, 54.3, 59.6],
    'n_i': [17, 23, 23, 35, 17, 1, 1],
    'u_i': [-3, -2, -1, 0, 1, 2, 3],
    'n_i*u_i': [-51, -46, -23, 0, 17, 2, 3],
    'n_i*u_i^2': [153, 92, 23, 0, 17, 4, 9],
    'n_i*u_i^3': [-459, -184, -23, 0, 17, 8, 27],
    'n_i*u_i^4': [1377, 368, 23, 0, 17, 16, 81],
    'n_i(u_i+1)^4': [272, 23, 0, 35, 272, 81, 256]
}

data_d = {
    'i': [1, 2, 3, 4, 5, 6, 7],
    'x_i': [13.65, 14.8, 15.95, 17.05, 18.25, 19.2, 20.55],
    'n_i': [23, 31, 19, 11, 17, 10, 6],
    'u_i': [-3, -2, -1, 0, 1, 2, 3],
    'n_i*u_i': [-69, -62, -19, 0, 17, 20, 18],
    'n_i*u_i^2': [207, 124, 19, 0, 17, 40, 54],
    'n_i*u_i^3': [-621, -248, -19, 0, 17, 80, 162],
    'n_i*u_i^4': [1863, 496, 19, 0, 17, 160, 486],
    'n_i(u_i+1)^4': [368, 31, 0, 11, 272, 810, 1536]
}

u_df_d = pd.DataFrame(data_d)
u_df_l = pd.DataFrame(data_l)
display(u_df_l)
display(u_df_d)

,i,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i(u_i+1)^4
0,1,36.20,17,-3,-51,153,-459,1377,272
1,2,39.85,23,-2,-46,92,-184,368,23
2,3,43.60,23,-1,-23,23,-23,23,0
3,4,47.00,35,0,0,0,0,0,35
4,5,49.85,17,1,17,17,17,17,272
5,6,54.30,1,2,2,4,8,16,81
6,7,59.60,1,3,3,9,27,81,256


,i,x_i,n_i,u_i,n_i*u_i,n_i*u_i^2,n_i*u_i^3,n_i*u_i^4,n_i(u_i+1)^4
0,1,13.65,23,-3,-69,207,-621,1863,368
1,2,14.80,31,-2,-62,124,-248,496,31
2,3,15.95,19,-1,-19,19,-19,19,0
3,4,17.05,11,0,0,0,0,0,11
4,5,18.25,17,1,17,17,17,17,272
5,6,19.20,10,2,20,40,80,160,810
6,7,20.55,6,3,18,54,162,486,1536


In [ ]:
def calculate_u_B(df):
  u_B = 0
  u_2 = 0
  for idx, row in df.iterrows():
    u_B += row['n_i*u_i']
    u_2 += row['n_i*u_i^2']
  return u_B/N, u_2/N

u_B_lenght, u_2_lenght = calculate_u_B(u_df_l)
u_B_depth, u_2_depth = calculate_u_B(u_df_d)
print('u_B_lenght =', round(u_B_lenght, 2), ' u_B_depth =', round(u_B_depth, 2))
print('u^2_lenght =', round(u_2_lenght, 2), ' u^2_depth =', round(u_2_depth, 2))
sigma_lenght = (u_2_lenght - u_B_lenght**2)**0.5
sigma_depth = (u_2_depth - u_B_depth**2)**0.5
print('sigma_lenght =', round(sigma_lenght, 2), ' sigma_depth =', round(sigma_depth, 2))

u_B_lenght = -0.84  u_B_depth = -0.81
u^2_lenght = 2.55  u^2_depth = 3.94
sigma_lenght = 1.36  sigma_depth = 1.81


In [ ]:
sum_uv = 0
for i in range(K):
    for j in range(K):
        n_ij = freq_matrix[i, j]
        if n_ij > 0:
            contribution = n_ij * u_df_l.loc[i]['u_i'] * u_df_d.loc[j]['u_i']
            sum_uv += contribution
# print(f"\nСумма ∑∑ n_ij * x_i * y_j = {sum_uv:.2f}")
r_u = (sum_uv - N * u_B_lenght * u_B_depth) / (N * sigma_lenght * sigma_depth)
print('Выборочный коэффициент корреляции, рассчитанный с помощью условных вариант: ', round(r, 2))

Выборочный коэффициент корреляции, рассчитанный с помощью условных вариант:  -0.56


Результаты совпадают.  Полученное значение коэфиициента r = -0.56 показывает умеренную отрицательную связь. Это означает, что с увеличением одной переменной другая переменная уменьшается, что подтверждает визуальный вывод, сделанный ранее по корреляционной таблице.

5. Построим доверительный интервал для коэффициента корреляции при уровне значимости $γ∈{0.95,0.99}$

In [ ]:
def conf_interv_for_r(z, sigma_z, lambda_gamma):
  z_1 = z - lambda_gamma*sigma_z
  z_2 = z + lambda_gamma*sigma_z
  print('(', round((math.e**(2*z_1)-1)/(math.e**(2*z_1)+1), 2), round((math.e**(2*z_2)-1)/(math.e**(2*z_2)+1), 2), ')')

In [ ]:
z = 0.5 * math.log((1+r)/(1-r), math.e)
print('z =', round(z, 2))
sigma_z = 1/(math.sqrt(N-3))
print('sigma_z =', round(sigma_z, 2))
lambda_gamma_0_475 = 1.96
lambda_gamma_0_495 = 2.58
print('Доверительный интервал для коэффициента корреляции при уровне значимости  γ = 0.95')

conf_interv_for_r(z, sigma_z, lambda_gamma_0_475)
print('Доверительный интервал для коэффициента корреляции при уровне значимости  γ = 0.99')
conf_interv_for_r(z, sigma_z, lambda_gamma_0_495)

z = -0.63
sigma_z = 0.09
Доверительный интервал для коэффициента корреляции при уровне значимости  γ = 0.95
( -0.67 -0.42 )
Доверительный интервал для коэффициента корреляции при уровне значимости  γ = 0.99
( -0.7 -0.37 )


При увеличении доверительной вероятности γ с 0.95 до 0.99 интервал расширяется с 0.25 до 0.35

6. Проверим статистическую гипотезу о равенстве коэффициента корреляции нулю при заданном уровне значимости α=0.05.

In [ ]:
T = (r*math.sqrt(N-2))/(1-r**2)
print('T =', round(T, 2))
t_115 = 1.98
print('t_крит =', t_115)

T = -8.73
t_крит = 1.98


$|T_{набл}|$ = 8.73 > 1.98 = $t_{крит}$

Значит основная гипотеза $H_0$: $r_{xy} = 0 $ - не согласована с выборочными данными и должна быть отвергнута. А это означает, что выборочный коэффициент вариации $\bar{r}_{xy}$ значимо отличается от нуля.

## Выводы

Построенная корреляционная таблица демонстрирует обратную зависимость между глубиной и длиной рыбы: с увеличением глубины длина уменьшается.

Выборочный коэффициент корреляции, рассчитанный двумя способами (по стандартной формуле и методом условных вариант), показал идентичный результат: $r_{xy} ≈−0.56$. Это свидетельствует об умеренной обратной заметной связи.

С надежностью γ=0.95 истинное значение коэффициента корреляции в генеральной совокупности находится в пределах (−0.68;−0.42), а с надежностью γ=0.99 – в пределах (−0.70;−0.37).

Наблюдаемое значение случайной величины $|T_{набл}|$   значительно превышает t-критическое. Это даёт основание отвергнуть нулевую гипотезу о равенстве коэффициента корреляции нулю на уровне значимости α=0.05.